### DISCUSSION TITLE
`Cell::as_slice_of_cells()` replaces `windows_mut` and eliminates bounds checks

### DISCUSSION THREAD

**Discussion by Shnatsel** [etemesi254/zune-image#59] [Ideas]
We've been having trouble with bounds checks in the RLE hot loop in `zune-inflate`, and couldn't eliminate them because the slices we wanted to operate on were overlapping. Since there's no `windows_mut` in Rust, we could not use iterators to get rid of bounds checks.

Following the release of [my article](https://shnatsel.medium.com/how-to-avoid-bounds-checks-in-rust-without-unsafe-f65e618b4c1e), a [PR](https://github.com/rust-lang/rust/pull/106889/files#diff-e8ccaf64ce21f955ccebef33b52158631493a6f0966815a2ebc142d7cd2b5e06R785) has been merged into the standard library explaining how to work around this. 

You can see it in action on the samples from the article [here](https://old.reddit.com/r/rust/comments/10edmjf/how_to_avoid_bounds_checks_in_rust_without_unsafe/j4u91of/?context=3).

**Comment by etemesi254** [comment]
On it

Lemme take a look this weekend .

Hope it helps in the benchmarks.

Also loved the article, (esp the anti-patterns, fell prey to some)


Will update this very soon

**Comment by etemesi254** [comment]
A little bit complicated, needs some comments around it
But does the job https://godbolt.org/z/Y4e4qqhaE
```rust
#[inline(always)]
pub fn copy_rep_matches_new(dest: &mut [u8], offset: usize, dest_offset: usize, length: usize)
{
    let diff = dest_offset - offset + 1;

    for window in Cell::from_mut(&mut dest[offset..offset+length+1])
        .as_slice_of_cells()
        .windows(diff)
    {
        window.last().unwrap().set(window[0].get());
    }
}
```

**Reply by Shnatsel** [reply]
This is copying one byte at a time. I know we also had a loop that was copying in batches, but still had bounds check per batch.

Was the batching purely to reduce the amount of bounds checks? If not, does it make sense to convert that loop to use `slice_of_cells` as well?

**Reply by etemesi254** [reply]
Batch loop js a lot harder to figure out.

May be possible but I'm not sure how.

**Reply by etemesi254** [reply]
The batching was to remove the number of copies we needed to do to finish a copy depending on the situation.

With this it becomes difficult, batch copy moves in batches and `windows()` can't move in batches, (always moves by one byte)

**Reply by Shnatsel** [reply]
Well, you could definitely eliminate several branches from `fixed_copy_within` just by using `slice_of_cells` + `windows` instead of `loop` + `break`. Although you'll probably have to hand-roll `fixed_copy_within` instead of just using `slice::copy_within`.

**Reply by etemesi254** [reply]
A small poc would be nice, as I can't see how that works.

The compiler does two bounds check there, one to check end is greater than start, and one to check end is in bounds.

Number of instructions to beat is 4

**Reply by Shnatsel** [reply]
Something like this? https://github.com/Shnatsel/zune-image/commit/5c88389663ceca0e6b1d6390d038165187eff495

But you're right, this really hinges on whether the compiler will load every byte individually when they're all behind a `Cell` or load them in bulk. 

My hope is that copying to an intermediate buffer will be desugared into a load and store into a 128-bit register. But if that's not the case, then this approach isn't viable.

**Reply by Shnatsel** [reply]
You were right, using `Cell`s in the overlapping bulk copy loop resulted in a massive performance drop. I'm going to abandon this line of inquiry.

**Comment by etemesi254** [comment]
Btw this Cells::as_slice is goood.


Kills bounds checks in png filterings


I.e this 

```Rust
pub fn handle_avg(
    prev_row: &[u8], raw: &[u8], current: &mut [u8], components: usize, _use_sse4: bool
)
{
    if raw.len() < components || current.len() < components
    {
        return;
    }


    // no simd, so just do it the old fashioned way

    // handle leftmost byte explicitly
    for i in 0..components
    {
        current[i] = raw[i].wrapping_add(prev_row[i] >> 1);
    }
    // raw length is one row,so always keep it in check
    let end = current.len().min(raw.len()).min(prev_row.len());

    if components > 8
    {
        // optimizer hint to tell the compiler that we don't see this ever happening
        return;
    }
    for i in components..end
    {
        let a = u16::from(current[i - components]);
        let b = u16::from(prev_row[i]);

        let c = (((a + b) >> 1) & 0xFF) as u8;

        current[i] = raw[i].wrapping_add(c);
    }
}
```

And this
```Rust
use std::cell::Cell;

pub fn handle_avg(
    prev_row: &[u8], raw: &[u8], current: &mut [u8], components: usize, _use_sse4: bool
)
{
    if raw.len() < components || current.len() < components
    {
        return;
    }
    // handle leftmost byte explicitly
    for i in 0..components
    {
        current[i] = raw[i].wrapping_add(prev_row[i] >> 1);
    }
    // raw length is one row,so always keep it in check
    let end = current.len().min(raw.len()).min(prev_row.len());

    if components > 8
    {
        // optimizer hint to tell the compiler that we don't see this ever happening
        return;
    }
    let current_as_cells = Cell::from_mut(&mut current[..end])
        .as_slice_of_cells()
        .windows(components + 1);

    for ((current_window, byte), r) in current_as_cells
        .zip(&prev_row[components..])
        .zip(&raw[components..])
    {
        let a = u16::from(current_window.first().unwrap().get());
        let b = u16::from(*byte);

        let c = (((a + b) >> 1) & 0xFF) as u8;

        current_window.last().unwrap().set(r.wrapping_add(c));
    }   
}
```
Are semantically equivalent,but the latter doesn't have bounds check. My only problem is that the latter is a code bloat since it generates sse code for paths greater than 8, the only one I can think of is 16 bit RGBA images(4 (colorspace)* 2(number of bytes)), I should probably add a benchmark for that.

**Comment by etemesi254** [comment]
Sadly doesn't help, e.g benchmark numbers for 16 bit png unfiltering

|decoder|speed|
|-----------|---------|
|zune-png|672 ms|
|image-rs/png-decoder|692 ms|
|spng | 711 ms|

If i add a SSE path for 16 bit unfiltering,we have
|decoder|speed|
|------------|----------|
|zune-png|  303.14 ms|
